src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -60,15 +64,35 @@ object RockSlicer {
       !block.faces.exists(_.processorJoint)
     }
 
-    // Process processor blocks before continueing with remaining blocks
-    // Collect all blocks that contain processor joints from all nodes
-    val allProcessorBlocks = processorBlocks.collect()
+    // Process processor blocks before continuing with remaining blocks
     // Search blocks for matching processor joints
-    if (allProcessorBlocks.length > 0) {
-      val reconstructedBlocks = mergeBlocks(allProcessorBlocks, Seq.empty[Block], globalOrigin)
+    if (processorBlocks.count() > 0) {
+      // Create pair RDD that contains initially all the blocks on each partion as the keys and
+      // an empty sequence of blocks as the values. The keys represent blocks that are yet to be merged,
+      // the values are blocks that have already been merged.
+      val treeReduceBlockPairsRDD = processorBlocks.map{ blocks => (Seq(blocks), Seq.empty[Block])}
+      val (allOrphanBlocks, allReconstructedBlocks) = treeReduceBlockPairsRDD.treeReduce({
+        case ((toMerge1, merged1), (toMerge2, merged2)) =>
+          // All blocks that share a processor joint are merged/reconstructed. Blocks that do not share
+          // a processor joint with any of the blocks that are to be merged are "orphan" blocks.
+          val (treeReconBlocks, treeOrphanBlocks) = mergeBlocks(toMerge1 ++ toMerge2, Seq.empty[Block],
+                                                                Seq.empty[Block])
+
+          // Merged/reconstructed blocks are partitioned based on whether they still contain processor joints
+          val (treeProcessorBlocks, treeRealBlocks) = treeReconBlocks.partition { block =>
+            block.faces.exists(_.processorJoint)
+          }
+
+          // Orphan blocks and merged/reconstructed blocks that still contain processor joints are returned
+          // as the keys to the pair RDD. These will be merged with the next level of the tree. Blocks that
+          // have no remaining processor joints are returned as the values in the pair RDD.
+          (treeProcessorBlocks ++ treeOrphanBlocks, treeRealBlocks ++ merged1 ++ merged2)
+      }, math.ceil(math.log(arguments.numProcessors)/math.log(2)).toInt)
```

Comment:
Since it took us a little while to come up with this formula, a brief comment explaining where this quantity came from would be useful. :)